File: impl/csharp/Handshake.cs
// Open reference implementation of the N-PAMP wire format (draft-bubblefish-npamp-00).
// Handshake binding layer (binding spec/10): transcript construction, Finished MAC,
// CertVerify signature. No proprietary methods, parameters, or weights. Big-endian throughout.
//
// SHA-256 + HMAC come from System.Security.Cryptography (the .NET BCL). Ed25519 is NOT in
// the BCL, so it comes from BouncyCastle.Cryptography (Org.BouncyCastle.Crypto.Signers.
// Ed25519Signer, pure Ed25519 per RFC 8032). Kept in its OWN file so the BCL-only AES-GCM
// KAT build (Npamp.cs alone) never references BouncyCastle.
#nullable enable
using System;
using System.Collections.Generic;
using System.Security.Cryptography;
using System.Text;
using Org.BouncyCastle.Crypto.Parameters;
using Org.BouncyCastle.Crypto.Signers;

namespace Sh.Bubblefish.Npamp;

/// <summary>
/// N-PAMP draft-00 handshake binding layer (binding spec/10): the transcript hash
/// (section 3), the Finished MAC (section 6.2; RFC 8446 section 4.4.4), and the
/// CertVerify signature (section 6.1; RFC 8446 section 4.4.3 structure; Ed25519 per
/// RFC 8032). Big-endian throughout.
/// </summary>
public static class Handshake
{
    // -- Constants ---------------------------------------------------------

    /// <summary>Handshake frame types (binding spec/10 section 1), carried on the control channel.</summary>
    public const int FrameClientHello = 0x0100, FrameServerHello = 0x0101,
        FrameServerAuth = 0x0102, FrameClientAuth = 0x0103;

    /// <summary>CertVerify role context strings (binding spec/10 section 6.1).</summary>
    public const string ContextServerCertVerify = "N-PAMP draft-00, server CertificateVerify";
    public const string ContextClientCertVerify = "N-PAMP draft-00, client CertificateVerify";

    // -- Big-endian helpers ------------------------------------------------

    private static int GetU16Be(byte[] buf, int off) => ((buf[off] & 0xFF) << 8) | (buf[off + 1] & 0xFF);

    // -- Transcript (binding spec/10 section 3) ----------------------------

    /// <summary>
    /// Accumulates the draft-00 handshake transcript (binding spec/10 section 3) and
    /// hashes it at a cut point. Absorption granularity is per-TLV: <see cref="AddFrameType"/>
    /// appends the 2-octet frame type ONLY (NOT the rest of the 36-octet frame header — the
    /// spec section 3 / 7.1 divergence from RFC 8446 section 4.4.1); <see cref="AddTLV"/>
    /// appends Type(2 BE) || Length(2 BE) || Value. A transcript point = the hash over all
    /// bytes absorbed so far (SHA-256 at Standard, SHA-384 at High/Sovereign).
    /// </summary>
    public sealed class Transcript
    {
        private readonly List<byte> _buf = new();

        /// <summary>Appends the frame type as exactly 2 octets big-endian.</summary>
        public void AddFrameType(int ft)
        {
            _buf.Add((byte)((ft >> 8) & 0xFF));
            _buf.Add((byte)(ft & 0xFF));
        }

        /// <summary>Appends one TLV: Type(2 BE) || Length(2 BE) || Value.</summary>
        public void AddTLV(int type, byte[] value)
        {
            _buf.Add((byte)((type >> 8) & 0xFF));
            _buf.Add((byte)(type & 0xFF));
            _buf.Add((byte)((value.Length >> 8) & 0xFF));
            _buf.Add((byte)(value.Length & 0xFF));
            _buf.AddRange(value);
        }

        /// <summary>Hashes every octet absorbed so far (SHA-256 when <paramref name="standard"/>, else SHA-384).</summary>
        public byte[] Hash(bool standard)
        {
            byte[] data = _buf.ToArray();
            return standard ? SHA256.HashData(data) : SHA384.HashData(data);
        }
    }

    // -- Finished (binding spec/10 section 6.2; RFC 8446 section 4.4.4) -----

    /// <summary>
    /// Finished verify_data = HMAC(finished_key, transcript_hash) under the profile hash
    /// (HMAC-SHA-256 at Standard, HMAC-SHA-384 at High/Sovereign).
    /// </summary>
    public static byte[] ComputeFinished(byte[] finishedKey, byte[] transcriptHash, bool standard)
    {
        return standard
            ? HMACSHA256.HashData(finishedKey, transcriptHash)
            : HMACSHA384.HashData(finishedKey, transcriptHash);
    }

    /// <summary>
    /// Recomputes the Finished MAC and constant-time-compares it to the received verify_data.
    /// <see cref="CryptographicOperations.FixedTimeEquals"/> is time-independent and returns
    /// false on a length mismatch.
    /// </summary>
    public static bool VerifyFinished(byte[] finishedKey, byte[] transcriptHash, byte[] verifyData, bool standard)
    {
        return CryptographicOperations.FixedTimeEquals(
            ComputeFinished(finishedKey, transcriptHash, standard), verifyData);
    }

    // -- Ed25519 key decoding (RFC 8032) -----------------------------------

    /// <summary>Builds an Ed25519 private key from its raw 32-octet seed (RFC 8032).</summary>
    public static Ed25519PrivateKeyParameters Ed25519PrivateKeyFromSeed(byte[] seed)
    {
        return new Ed25519PrivateKeyParameters(seed, 0);
    }

    /// <summary>Builds an Ed25519 public key from its raw 32-octet encoding (RFC 8032 section 5.1.2).</summary>
    public static Ed25519PublicKeyParameters Ed25519PublicKeyFromRaw(byte[] raw)
    {
        if (raw.Length != 32)
        {
            throw new ArgumentException("ed25519 public key must be 32 octets, got " + raw.Length);
        }
        return new Ed25519PublicKeyParameters(raw, 0);
    }

    // -- CertVerify (binding spec/10 section 6.1; RFC 8446 section 4.4.3) ---

    /// <summary>
    /// The section 6.1 signing input: 64 octets of 0x20, the role context string, a 0x00
    /// separator, then the transcript hash — TLS-1.3-style domain separation
    /// (RFC 8446 section 4.4.3).
    /// </summary>
    public static byte[] CertVerifySigningInput(bool isServer, byte[] transcriptHash)
    {
        byte[] ctx = Encoding.ASCII.GetBytes(isServer ? ContextServerCertVerify : ContextClientCertVerify);
        byte[] outBuf = new byte[64 + ctx.Length + 1 + transcriptHash.Length];
        int p = 0;
        for (int i = 0; i < 64; i++)
        {
            outBuf[p++] = 0x20;
        }
        Array.Copy(ctx, 0, outBuf, p, ctx.Length);
        p += ctx.Length;
        outBuf[p++] = 0x00;
        Array.Copy(transcriptHash, 0, outBuf, p, transcriptHash.Length);
        return outBuf;
    }

    /// <summary>The CertVerify TLV value: u16(0x0807, Ed25519) || Ed25519(priv, signing_input).</summary>
    public static byte[] SignCertVerify(Ed25519PrivateKeyParameters privateKey, bool isServer, byte[] transcriptHash)
    {
        var signer = new Ed25519Signer();
        signer.Init(true, privateKey);
        byte[] msg = CertVerifySigningInput(isServer, transcriptHash);
        signer.BlockUpdate(msg, 0, msg.Length);
        byte[] sig = signer.GenerateSignature();
        byte[] outBuf = new byte[2 + sig.Length];
        outBuf[0] = (byte)((Npamp.SigEd25519 >> 8) & 0xFF);
        outBuf[1] = (byte)(Npamp.SigEd25519 & 0xFF);
        Array.Copy(sig, 0, outBuf, 2, sig.Length);
        return outBuf;
    }

    /// <summary>
    /// Checks a CertVerify TLV value against the signer's public key, role, and transcript
    /// hash. Rejects a non-Ed25519 scheme, a wrong-length (non-64-octet) signature, a
    /// role/context mismatch, or a wrong transcript.
    /// </summary>
    public static bool VerifyCertVerify(Ed25519PublicKeyParameters publicKey, bool isServer, byte[] transcriptHash, byte[] value)
    {
        if (value.Length < 2 || GetU16Be(value, 0) != Npamp.SigEd25519)
        {
            return false;
        }
        byte[] sig = new byte[value.Length - 2];
        Array.Copy(value, 2, sig, 0, sig.Length);
        if (sig.Length != 64) // Ed25519 signatures are exactly 64 octets (RFC 8032 section 5.1.6)
        {
            return false;
        }
        try
        {
            var verifier = new Ed25519Signer();
            verifier.Init(false, publicKey);
            byte[] msg = CertVerifySigningInput(isServer, transcriptHash);
            verifier.BlockUpdate(msg, 0, msg.Length);
            return verifier.VerifySignature(sig);
        }
        catch (Exception)
        {
            return false;
        }
    }
}

File: impl/go/handshake.go
package npamp

import (
	"crypto/ed25519"
	"crypto/hmac"
	"encoding/binary"
	"errors"
	"fmt"
)

// Handshake frame types on the Control channel (spec/10 section 1.1). The
// handshake is a four-frame 1.5-RTT exchange with mutual authentication;
// there is no separate Finished frame — the Finished MAC is a TLV inside each
// AUTH frame. These occupy the channel-specific space (>= ChannelSpecificBase)
// and are frozen for draft-01.
const (
	FrameClientHello FrameType = 0x0100 // cleartext
	FrameServerHello FrameType = 0x0101 // cleartext
	FrameServerAuth  FrameType = 0x0102 // AEAD-sealed under the s-hs handshake key
	FrameClientAuth  FrameType = 0x0103 // AEAD-sealed under the c-hs handshake key
)

// Role identifies which side of the handshake produced an authentication
// message. The CertVerify context string differs per role, so a server
// CertVerify is unusable as a client one (spec/10 section 6.1).
type Role uint8

const (
	RoleClient Role = 1
	RoleServer Role = 2
)

var (
	ErrHandshakeTLVOrder   = errors.New("npamp: handshake TLVs missing, extra, or out of order")
	ErrHandshakeTLVValue   = errors.New("npamp: handshake TLV value has an invalid size")
	ErrInvalidRole         = errors.New("npamp: role must be RoleClient or RoleServer")
	ErrCertVerifyScheme    = errors.New("npamp: CertVerify signature scheme was not negotiated")
	ErrCertVerifySignature = errors.New("npamp: CertVerify signature verification failed")
	ErrCertVerifyKeySize   = errors.New("npamp: identity key is not an Ed25519 public key")
	ErrFinishedMismatch    = errors.New("npamp: Finished verify_data mismatch")
)

// certVerifyContext returns the role's CertVerify context string
// (spec/10 section 6.1), or "" for an invalid role.
func (r Role) certVerifyContext() string {
	switch r {
	case RoleClient:
		return "N-PAMP/2, client CertificateVerify"
	case RoleServer:
		return "N-PAMP/2, server CertificateVerify"
	default:
		return ""
	}
}

// ClientHello is the first handshake flight (frame type 0x0100, cleartext):
// ProfileOffer, KEMOffer, SigOffer, AEADOffer, KEMShare, in that order
// (spec/10 section 1).
type ClientHello struct {
	ProfileOffer []Profile
	KEMOffer     []KEMID
	SigOffer     []SigID
	AEADOffer    []AEADID
	KEMShare     []byte
}

// ServerHello is the second handshake flight (frame type 0x0101, cleartext):
// ProfileSelect, KEMSelect, SigSelect, AEADSelect, KEMCiphertext, in that
// order (spec/10 section 1).
type ServerHello struct {
	ProfileSelect Profile
	KEMSelect     KEMID
	SigSelect     SigID
	AEADSelect    AEADID
	KEMCiphertext []byte
}

// AuthMessage is the SERVER_AUTH (0x0102) / CLIENT_AUTH (0x0103) plaintext:
// exactly three TLVs in order — IdentityKey, CertVerify, Finished
// (spec/10 sections 1 and 6.4). The frames carrying it are AEAD-sealed under
// the per-direction handshake traffic key.
type AuthMessage struct {
	IdentityKey []byte // Ed25519 public key (32 octets at Standard)
	CertVerify  []byte // SignatureScheme uint16 || signature
	Finished    []byte // HashLen-octet HMAC verify_data
}

// u16ListBytes encodes a list of 16-bit code points big-endian.
func u16ListBytes[T ~uint16](ids []T) []byte {
	out := make([]byte, 0, 2*len(ids))
	for _, id := range ids {
		out = binary.BigEndian.AppendUint16(out, uint16(id))
	}
	return out
}

// parseU16List decodes a big-endian list of 16-bit code points; the value must
// be non-empty and of even length.
func parseU16List[T ~uint16](v []byte, name string) ([]T, error) {
	if len(v) == 0 || len(v)%2 != 0 {
		return nil, fmt.Errorf("%w: %s is %d octets", ErrHandshakeTLVValue, name, len(v))
	}
	out := make([]T, 0, len(v)/2)
	for i := 0; i < len(v); i += 2 {
		out = append(out, T(binary.BigEndian.Uint16(v[i:i+2])))
	}
	return out, nil
}

// requireTLVs checks that tlvs is exactly the expected types in the expected
// order (the handshake fixes both, spec/10 section 1).
func requireTLVs(tlvs []TLV, want []TLVType) error {
	if len(tlvs) != len(want) {
		return fmt.Errorf("%w: got %d TLVs, want %d", ErrHandshakeTLVOrder, len(tlvs), len(want))
	}
	for i, w := range want {
		if tlvs[i].Type != w {
			return fmt.Errorf("%w: position %d is 0x%02x, want 0x%02x", ErrHandshakeTLVOrder, i, uint16(tlvs[i].Type), uint16(w))
		}
	}
	return nil
}

// TLVs returns the flight's TLVs in transcript/wire order.
func (m *ClientHello) TLVs() []TLV {
	profiles := make([]byte, len(m.ProfileOffer))
	for i, p := range m.ProfileOffer {
		profiles[i] = byte(p)
	}
	return []TLV{
		{Type: TLVProfileOffer, Value: profiles},
		{Type: TLVKEMOffer, Value: u16ListBytes(m.KEMOffer)},
		{Type: TLVSigOffer, Value: u16ListBytes(m.SigOffer)},
		{Type: TLVAEADOffer, Value: u16ListBytes(m.AEADOffer)},
		{Type: TLVKEMShare, Value: m.KEMShare},
	}
}

// Encode returns the CLIENT_HELLO frame payload: the TLVs concatenated in
// canonical form.
func (m *ClientHello) Encode() []byte {
	var out []byte
	for _, v := range m.TLVs() {
		out = v.Encode(out)
	}
	return out
}

// DecodeClientHello parses a CLIENT_HELLO frame payload, enforcing the exact
// TLV set and order of spec/10 section 1.
func DecodeClientHello(payload []byte) (*ClientHello, error) {
	tlvs, err := DecodeTLVs(payload)
	if err != nil {
		return nil, err
	}
	if err := requireTLVs(tlvs, []TLVType{TLVProfileOffer, TLVKEMOffer, TLVSigOffer, TLVAEADOffer, TLVKEMShare}); err != nil {
		return nil, err
	}
	if len(tlvs[0].Value) == 0 {
		return nil, fmt.Errorf("%w: ProfileOffer is empty", ErrHandshakeTLVValue)
	}
	profiles := make([]Profile, len(tlvs[0].Value))
	for i, b := range tlvs[0].Value {
		profiles[i] = Profile(b)
	}
	kems, err := parseU16List[KEMID](tlvs[1].Value, "KEMOffer")
	if err != nil {
		return nil, err
	}
	sigs, err := parseU16List[SigID](tlvs[2].Value, "SigOffer")
	if err != nil {
		return nil, err
	}
	aeads, err := parseU16List[AEADID](tlvs[3].Value, "AEADOffer")
	if err != nil {
		return nil, err
	}
	if len(tlvs[4].Value) == 0 {
		return nil, fmt.Errorf("%w: KEMShare is empty", ErrHandshakeTLVValue)
	}
	return &ClientHello{
		ProfileOffer: profiles,
		KEMOffer:     kems,
		SigOffer:     sigs,
		AEADOffer:    aeads,
		KEMShare:     tlvs[4].Value,
	}, nil
}

// TLVs returns the flight's TLVs in transcript/wire order.
func (m *ServerHello) TLVs() []TLV {
	return []TLV{
		{Type: TLVProfileSelect, Value: []byte{byte(m.ProfileSelect)}},
		{Type: TLVKEMSelect, Value: binary.BigEndian.AppendUint16(nil, uint16(m.KEMSelect))},
		{Type: TLVSigSelect, Value: binary.BigEndian.AppendUint16(nil, uint16(m.SigSelect))},
		{Type: TLVAEADSelect, Value: binary.BigEndian.AppendUint16(nil, uint16(m.AEADSelect))},
		{Type: TLVKEMCiphertext, Value: m.KEMCiphertext},
	}
}

// Encode returns the SERVER_HELLO frame payload: the TLVs concatenated in
// canonical form.
func (m *ServerHello) Encode() []byte {
	var out []byte
	for _, v := range m.TLVs() {
		out = v.Encode(out)
	}
	return out
}

// DecodeServerHello parses a SERVER_HELLO frame payload, enforcing the exact
// TLV set and order of spec/10 section 1 and the fixed select-value sizes
// (ProfileSelect 1 octet; KEM/Sig/AEADSelect 2 octets each).
func DecodeServerHello(payload []byte) (*ServerHello, error) {
	tlvs, err := DecodeTLVs(payload)
	if err != nil {
		return nil, err
	}
	if err := requireTLVs(tlvs, []TLVType{TLVProfileSelect, TLVKEMSelect, TLVSigSelect, TLVAEADSelect, TLVKEMCiphertext}); err != nil {
		return nil, err
	}
	if len(tlvs[0].Value) != 1 {
		return nil, fmt.Errorf("%w: ProfileSelect is %d octets, want 1", ErrHandshakeTLVValue, len(tlvs[0].Value))
	}
	for i, name := range []string{"", "KEMSelect", "SigSelect", "AEADSelect"} {
		if i == 0 {
			continue
		}
		if len(tlvs[i].Value) != 2 {
			return nil, fmt.Errorf("%w: %s is %d octets, want 2", ErrHandshakeTLVValue, name, len(tlvs[i].Value))
		}
	}
	if len(tlvs[4].Value) == 0 {
		return nil, fmt.Errorf("%w: KEMCiphertext is empty", ErrHandshakeTLVValue)
	}
	return &ServerHello{
		ProfileSelect: Profile(tlvs[0].Value[0]),
		KEMSelect:     KEMID(binary.BigEndian.Uint16(tlvs[1].Value)),
		SigSelect:     SigID(binary.BigEndian.Uint16(tlvs[2].Value)),
		AEADSelect:    AEADID(binary.BigEndian.Uint16(tlvs[3].Value)),
		KEMCiphertext: tlvs[4].Value,
	}, nil
}

// TLVs returns the AUTH message's TLVs in transcript/wire order.
func (m *AuthMessage) TLVs() []TLV {
	return []TLV{
		{Type: TLVIdentityKey, Value: m.IdentityKey},
		{Type: TLVCertVerify, Value: m.CertVerify},
		{Type: TLVFinished, Value: m.Finished},
	}
}

// Encode returns the AUTH frame plaintext (sealed by the record layer): the
// three TLVs concatenated in canonical form.
func (m *AuthMessage) Encode() []byte {
	var out []byte
	for _, v := range m.TLVs() {
		out = v.Encode(out)
	}
	return out
}

// DecodeAuthMessage parses an opened SERVER_AUTH / CLIENT_AUTH plaintext.
// Exactly three TLVs in order (IdentityKey, CertVerify, Finished) are required
// (spec/10 section 6.4).
func DecodeAuthMessage(payload []byte) (*AuthMessage, error) {
	tlvs, err := DecodeTLVs(payload)
	if err != nil {
		return nil, err
	}
	if err := requireTLVs(tlvs, []TLVType{TLVIdentityKey, TLVCertVerify, TLVFinished}); err != nil {
		return nil, err
	}
	for i, name := range []string{"IdentityKey", "CertVerify", "Finished"} {
		if len(tlvs[i].Value) == 0 {
			return nil, fmt.Errorf("%w: %s is empty", ErrHandshakeTLVValue, name)
		}
	}
	return &AuthMessage{
		IdentityKey: tlvs[0].Value,
		CertVerify:  tlvs[1].Value,
		Finished:    tlvs[2].Value,
	}, nil
}

// CertVerifySigningInput builds the RFC 8446 section 4.4.3-style input the
// CertVerify signature covers (spec/10 section 6.1):
//
//	0x20 x 64 || context || 0x00 || transcript_hash
//
// where transcript_hash is TH_sId (server) / TH_cId (client) — the transcript
// through the signer's own IdentityKey, before its own CertVerify.
func CertVerifySigningInput(role Role, transcriptHash []byte) ([]byte, error) {
	ctx := role.certVerifyContext()
	if ctx == "" {
		return nil, ErrInvalidRole
	}
	out := make([]byte, 0, 64+len(ctx)+1+len(transcriptHash))
	for range 64 {
		out = append(out, 0x20)
	}
	out = append(out, ctx...)
	out = append(out, 0x00)
	return append(out, transcriptHash...), nil
}

// SignCertVerify produces the TLV 0x0A value for the given role:
// SignatureScheme uint16 (Ed25519 = 0x0807) || Ed25519 signature over
// CertVerifySigningInput (spec/10 section 6.1). Standard profile signs with
// Ed25519 only.
func SignCertVerify(priv ed25519.PrivateKey, role Role, transcriptHash []byte) ([]byte, error) {
	input, err := CertVerifySigningInput(role, transcriptHash)
	if err != nil {
		return nil, err
	}
	sig := ed25519.Sign(priv, input)
	out := binary.BigEndian.AppendUint16(make([]byte, 0, 2+len(sig)), uint16(SigEd25519))
	return append(out, sig...), nil
}

// VerifyCertVerify checks a TLV 0x0A value against the peer's identity key,
// role, and transcript hash. It rejects a signature scheme other than the
// negotiated Ed25519 (0x0807), and — because the role selects the context
// string — a server CertVerify presented as a client one (spec/10 section
// 6.1).
func VerifyCertVerify(pub []byte, role Role, transcriptHash, certVerifyValue []byte) error {
	if len(pub) != ed25519.PublicKeySize {
		return ErrCertVerifyKeySize
	}
	if len(certVerifyValue) != 2+ed25519.SignatureSize {
		return fmt.Errorf("%w: value is %d octets, want %d", ErrCertVerifySignature, len(certVerifyValue), 2+ed25519.SignatureSize)
	}
	if SigID(binary.BigEndian.Uint16(certVerifyValue[:2])) != SigEd25519 {
		return ErrCertVerifyScheme
	}
	input, err := CertVerifySigningInput(role, transcriptHash)
	if err != nil {
		return err
	}
	if !ed25519.Verify(ed25519.PublicKey(pub), input, certVerifyValue[2:]) {
		return ErrCertVerifySignature
	}
	return nil
}

// ComputeFinished produces the TLV 0x0B verify_data (spec/10 section 6.2, per
// RFC 8446 section 4.4.4):
//
//	verify_data = HMAC(finished_key, transcript_hash)
//
// with HMAC over the profile's KDF hash. transcript_hash is TH_sCV (server) /
// TH_cCV (client) — the transcript through the sender's own CertVerify,
// excluding its own Finished. finished_key comes from DeriveFinishedKey.
func ComputeFinished(finishedKey, transcriptHash []byte, p Profile) []byte {
	m := hmac.New(hashForProfile(p), finishedKey)
	m.Write(transcriptHash)
	return m.Sum(nil)
}

// VerifyFinished recomputes the Finished MAC and compares it to verifyData in
// constant time, returning ErrFinishedMismatch on any difference (spec/10
// section 6.2 requires constant-time verification and abort on mismatch).
func VerifyFinished(finishedKey, transcriptHash, verifyData []byte, p Profile) error {
	want := ComputeFinished(finishedKey, transcriptHash, p)
	if !hmac.Equal(want, verifyData) {
		return ErrFinishedMismatch
	}
	return nil
}
